vendor/bat-native-ads/src/bat/ads/internal/ad_targeting/processors/behavioral/bandits/epsilon_greedy_bandit_processor.cc

```diff
@@ -0,0 +1,151 @@
+
+#include "bat/ads/internal/ad_targeting/processors/behavioral/bandits/epsilon_greedy_bandit_processor.h"
+
+#include <algorithm>
+#include <vector>
+#include <utility>
+
+#include "base/strings/string_number_conversions.h"
+#include "bat/ads/internal/ads_client_helper.h"
+#include "bat/ads/internal/ad_targeting/data_types/behavioral/bandits/epsilon_greedy_bandit_arms.h"
+#include "bat/ads/internal/ad_targeting/resources/behavioral/bandits/epsilon_greedy_bandit_resource.h"
+#include "bat/ads/internal/ad_targeting/ad_targeting_util.h"
+#include "bat/ads/internal/logging.h"
+#include "bat/ads/pref_names.h"
+
+namespace ads {
+namespace ad_targeting {
+namespace processor {
+
+namespace {
+const double kArmDefaultValue = 1.0;
+const uint64_t kArmDefaultPulls = 0;
+}  // namespace
+
+EpsilonGreedyBandit::EpsilonGreedyBandit() {
+  InitializeArms();
+}
+
+EpsilonGreedyBandit::~EpsilonGreedyBandit() = default;
+
+void EpsilonGreedyBandit::Process(
+    const BanditFeedback& feedback) {
+  const std::string segment = GetParentSegment(feedback.segment);
+
+  switch (feedback.ad_event_type) {
+    case AdNotificationEventType::kTimedOut:
+    case AdNotificationEventType::kDismissed: {
+      UpdateArm(/* reward */ 0, segment);
+      break;
+    }
+    case AdNotificationEventType::kClicked: {
+      UpdateArm(/* reward */ 1, segment);
+      break;
+    }
+    case AdNotificationEventType::kViewed: {
+      NOTREACHED();
+      break;
+    }
+  }
+
+  BLOG(1, "Processed ad event");
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+EpsilonGreedyBanditArmMap EpsilonGreedyBandit::MaybeAddOrResetArms(
+    const EpsilonGreedyBanditArmMap& arms) const {
+  EpsilonGreedyBanditArmMap updated_arms = arms;
+  for (const auto& segment : resource::kSegments) {
+    auto iter = updated_arms.find(segment);
+
+    if (iter == updated_arms.end()) {
+      EpsilonGreedyBanditArmInfo arm;
+      arm.value = kArmDefaultValue;
+      arm.pulls = kArmDefaultPulls;
+      updated_arms[segment] = arm;
+      BLOG(6, "Added arm for segment " << segment);
+      continue;
+    }
+
+    if (!iter->second.IsValid()) {
+      EpsilonGreedyBanditArmInfo new_arm;
+      new_arm.value = kArmDefaultValue;
+      new_arm.pulls = kArmDefaultPulls;
+      updated_arms[segment] = new_arm;
+      BLOG(6, "Did reset invalid arm for segment  " << segment);
+      continue;
+    }
+
+    BLOG(6, "Arm exists for segment " << segment);
+  }
+
+  return updated_arms;
+}
+
+EpsilonGreedyBanditArmMap EpsilonGreedyBandit::MaybeDeleteArms(
+    const EpsilonGreedyBanditArmMap& arms) const {
+  EpsilonGreedyBanditArmMap updated_arms = arms;
+  for (const auto& arm : updated_arms) {
+    auto iter = std::find(resource::kSegments.begin(),
+        resource::kSegments.end(), arm.first);
+
+    if (iter == resource::kSegments.end()) {
+      updated_arms.erase(arm.first);
+      BLOG(6, "Deleted arm for segment " << arm.first);
+    }
+  }
```

Comment:
Could this use `std::erase_if`?

Reply:
hm, we would have to make the log conditional so not saving much - wdyt?